File: common/src/main/java/com/jon/common/presets/PresetRepository.java
package com.jon.common.presets;

import androidx.room.Room;

import com.jon.common.CotApplication;
import com.jon.common.R;
import com.jon.common.utils.FileUtils;
import com.jon.common.utils.Protocol;

import org.apache.commons.collections4.ListUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.reactivex.Observable;

public class PresetRepository {
    private static final List<OutputPreset> UDP_DEFAULTS = new ArrayList<OutputPreset>() {{
        add(new OutputPreset(
                Protocol.UDP,
                CotApplication.getContext().getString(R.string.udpDefaultSa),
                CotApplication.getContext().getString(R.string.udpDefaultSaIp),
                Integer.parseInt(CotApplication.getContext().getString(R.string.udpDefaultSaPort))
        ));
    }};
    private static final List<OutputPreset> TCP_DEFAULTS = new ArrayList<OutputPreset>() {{
        add(new OutputPreset(
                Protocol.TCP,
                CotApplication.getContext().getString(R.string.tcpFreetakserver),
                CotApplication.getContext().getString(R.string.tcpFreetakserverIp),
                Integer.parseInt(CotApplication.getContext().getString(R.string.tcpFreetakserverPort))
        ));
    }};
    private static final List<OutputPreset> SSL_DEFAULTS = new ArrayList<OutputPreset>() {{
        add(new OutputPreset(
                Protocol.SSL,
                CotApplication.getContext().getString(R.string.sslTakserver),
                CotApplication.getContext().getString(R.string.sslTakserverIp),
                Integer.parseInt(CotApplication.getContext().getString(R.string.sslTakserverPort)),
                FileUtils.toByteArraySafe(R.raw.discord_client),
                "atakatak",
                FileUtils.toByteArraySafe(R.raw.discord_truststore),
                "atakatak"
        ));
    }};

    private static PresetRepository instance;
    private PresetDatabase database;
    private final Executor executor = Executors.newSingleThreadExecutor();

    synchronized public static PresetRepository getInstance() {
        if (instance == null) {
            instance = new PresetRepository();
        }
        return instance;
    }

    private PresetRepository() {
        database = Room.databaseBuilder(CotApplication.getContext(), PresetDatabase.class, PresetDatabase.FILENAME)
                .addMigrations(DatabaseMigrations.getAllMigrations())
                .build();
    }

    public void insertPreset(OutputPreset preset) {
        executor.execute(() -> database.presetDao().insert(preset));
    }

    public void deletePreset(OutputPreset preset) {
        executor.execute(() -> database.presetDao().delete(preset));
    }

    public void updatePreset(OutputPreset original, OutputPreset updated) {
        executor.execute(() -> database.presetDao().update(
                original.protocol.get(), original.address, original.port,
                updated.protocol.get(), updated.address, updated.port, updated.alias,
                updated.clientCert, updated.clientCertPassword, updated.trustStore, updated.trustStorePassword
        ));
    }

    public OutputPreset getPreset(Protocol protocol, String address, int port) {
        return database.presetDao().getPreset(protocol.get(), address, port);
    }

    /* Returns a list of all presets; meaning custom presets and defaults. */
    public Observable<List<OutputPreset>> getByProtocol(Protocol protocol) {
        /* First get a list of known default presets, which aren't stored in the database */
        final Observable<List<OutputPreset>> defaults = Observable.just(defaultsByProtocol(protocol));
        /* Then query the database to grab any user-entered presets */
        return getCustomByProtocol(protocol)
                /* Merge the two observables together to return a single observable containing a list of OutputPresets */
                .zipWith(defaults, (fetchedPresets, defaultPresets) -> ListUtils.union(defaultPresets, fetchedPresets));
    }

    /* Returns a list of only those presets entered by the user */
    public Observable<List<OutputPreset>> getCustomByProtocol(Protocol protocol) {
        return database.presetDao().getByProtocol(protocol.get());
    }

    public void deleteDatabase() {
        executor.execute(() -> database.presetDao().deleteAll());
    }

    public List<OutputPreset> defaultsByProtocol(Protocol protocol) {
        switch (protocol) {
            case SSL: return SSL_DEFAULTS;
            case TCP: return TCP_DEFAULTS;
            case UDP: return UDP_DEFAULTS;
            default: throw new IllegalArgumentException("Unknown protocol: " + protocol);
        }
    }
}

File: common/src/main/java/com/jon/common/repositories/PresetRepository.kt
package com.jon.common.repositories

import androidx.annotation.StringRes
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.jon.common.CotApplication
import com.jon.common.R
import com.jon.common.presets.OutputPreset
import com.jon.common.presets.PresetDatabase
import com.jon.common.utils.FileUtils
import com.jon.common.utils.Protocol
import java.util.concurrent.Executor
import java.util.concurrent.Executors


class PresetRepository private constructor() {
    private val database = PresetDatabase.build()
    private val executor: Executor = Executors.newSingleThreadExecutor()

    fun insertPreset(preset: OutputPreset) {
        executor.execute { database.presetDao().insert(preset) }
    }

    fun deletePreset(preset: OutputPreset) {
        executor.execute { database.presetDao().delete(preset) }
    }

    fun getPreset(protocol: Protocol, address: String, port: Int): OutputPreset? {
        return database.presetDao().getPreset(protocol.toString(), address, port)
    }

    /* Returns a list of only those presets entered by the user */
    fun getCustomByProtocol(protocol: Protocol): LiveData<List<OutputPreset>> {
        return database.presetDao().getByProtocol(protocol.toString())
    }

    fun deleteDatabase() {
        executor.execute { database.presetDao().deleteAll() }
    }

    fun defaultsByProtocol(protocol: Protocol): List<OutputPreset> {
        return when (protocol) {
            Protocol.SSL -> SSL_DEFAULTS
            Protocol.TCP -> TCP_DEFAULTS
            Protocol.UDP -> UDP_DEFAULTS
        }
    }

    companion object {
        private val instance = PresetRepository()
        fun getInstance() = instance

        private const val PASSWORD = "atakatak"
        private val UDP_DEFAULTS = listOf(
                OutputPreset(
                        Protocol.UDP,
                        getString(R.string.udp_default_sa),
                        getString(R.string.udp_default_sa_ip),
                        getString(R.string.udp_default_sa_port).toInt()
                )
        )
        private val TCP_DEFAULTS = listOf(
                OutputPreset(
                        Protocol.TCP,
                        getString(R.string.tcp_freetakserver_name),
                        getString(R.string.tcp_freetakserver_ip),
                        getString(R.string.tcp_freetakserver_port).toInt()
                )
        )
        private val SSL_DEFAULTS = listOf(
                OutputPreset(
                        Protocol.SSL,
                        getString(R.string.ssl_takserver_name),
                        getString(R.string.ssl_takserver_ip),
                        getString(R.string.ssl_takserver_port).toInt(),
                        FileUtils.rawResourceToByteArray(R.raw.discord_client),
                        PASSWORD,
                        FileUtils.rawResourceToByteArray(R.raw.discord_truststore),
                        PASSWORD
                )
        )

        private fun getString(@StringRes id: Int): String {
            return CotApplication.context.getString(id)
        }
    }
}
